File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    void solve(int idx, int n, vector<int>& ds, int target,
               vector<int>& candidates, vector<vector<int>>& ans) {
        if (idx == n) {
            if (target == 0) {
                ans.push_back(ds);
            }
            return;
        }

        // pick
        if (target >= candidates[idx]) {
            ds.push_back(candidates[idx]);
            solve(idx, n, ds, target - candidates[idx], candidates, ans);
            ds.pop_back();
        }

        // not pick
        solve(idx + 1, n, ds, target, candidates, ans);
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> ans;
        vector<int> ds;
        int n = candidates.size();
        solve(0, n, ds, target, candidates, ans);

        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    static void solve(int idx, int target, List<List<Integer>> res, List<Integer> ds, int[] cand) {
        if (idx == cand.length) {
            if (target == 0) {
                res.add(new ArrayList<>(ds));
            }
            return;
        }

        // pick
        if (target >= cand[idx]) {
            ds.add(cand[idx]);
            solve(idx, target - cand[idx], res, ds, cand);
            ds.remove(ds.size() - 1);
        }

        // not-pic
        solve(idx + 1, target, res, ds, cand);
    }

    public List<List<Integer>> combinationSum(int[] cand, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> ds = new ArrayList<>();

        solve(0, target, res, ds, cand);
        return res;
    }
}